### PR TITLE
Allow Dataptr_or_null method in ALTLIST classes

### PR DESCRIPTION
**Pull request by gaborcsardi**
This can be very useful to iterate over an ALTLIST
very quickly. For example, in an ALTLIST class
that implements a sublist, no copying is needed.

This method always returns a const pointer, so
it cannot be used to change the list.

### CODE REVIEW

File: src/main/altrep.c [hunk 1]
```diff
@@ -351,6 +351,10 @@ R_xlen_t /*attribute_hidden*/ ALTREP_TRUELENGTH(SEXP x) { return 0; }
 
 static R_INLINE void *ALTVEC_DATAPTR_EX(SEXP x, Rboolean writeable)
 {
+    /* protect VECSXP, it can only be changed via SET_VECTOR_ELT() */
```
Comment (lionel-) [inline src/main/altrep.c]:
```suggestion
    /* Disallow taking the `DATAPTR()` of an ALTLIST. This check could be
       moved to `DATAPTR()` to catch more faulty usages. */
```

File: src/main/altrep.c [hunk 2]
```diff
@@ -351,6 +351,10 @@ R_xlen_t /*attribute_hidden*/ ALTREP_TRUELENGTH(SEXP x) { return 0; }
...
+    if (TYPEOF(x) == VECSXP && writeable)
+        ALTREP_ERROR_IN_CLASS("ALTREP lists must not return a writeable DATAPTR", x);
```
Comment (lionel-) [inline src/main/altrep.c]:
```suggestion
        ALTREP_ERROR_IN_CLASS("cannot take a writeable DATAPTR of an ALTLIST", x);
```

**Review by lionel- — COMMENTED** [review]
It's probably better to merge the two commits in order to produce a single patch.

**Comment by gaborcsardi** [comment]
This has been merged.

**Pull request closed without merging**